clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -4298,6 +4296,39 @@ void handleFailure(Throwable throwable) {
                         }
                     }
                 }
+
+                @Override
+                boolean handleUnsupportedVersionException(UnsupportedVersionException exception) {
+                    if (supportsMaxTimestamp) {
+                        supportsMaxTimestamp = false;
+
+                        // fail any unsupported futures and remove partitions from the downgraded retry
+                        List<ListOffsetsTopic> topicsToRemove = new ArrayList<>();
+                        partitionsToQuery.stream().forEach(
+                            t -> {
+                                List<ListOffsetsPartition> partitionsToRemove = new ArrayList<>();
+                                t.partitions().stream()
+                                    .filter(p -> p.timestamp() == ListOffsetsRequest.MAX_TIMESTAMP)
+                                    .forEach(
+                                        p -> {
+                                            futures.get(new TopicPartition(t.name(), p.partitionIndex()))
+                                                .completeExceptionally(
+                                                    new UnsupportedVersionException(
+                                                        "Broker " + brokerId
+                                                            + " does not support MAX_TIMESTAMP offset spec"));
+                                            partitionsToRemove.add(p);
+
+                                        });
+                                t.partitions().removeAll(partitionsToRemove);
+                                if (t.partitions().isEmpty()) topicsToRemove.add(t);
+                            }
+                        );
+                        partitionsToQuery.removeAll(topicsToRemove);
+
+                        return !partitionsToQuery.isEmpty();
```

Comment:
Don't we need to not retry if they were no `MAX_TIMESTAMP` in the collection? In this case, it would mean that the `UnsupportedVersionException` is not retryable. This could happen if the admin client talk to a really old broker which does not support v1 for instance.

Reply:
good point, I've added a check for this.